File: src/2025-02-24-deepclaude-rust/java/exercises/practice/rest-api/src/main/java/RestApi.java
import org.json.JSONObject;
import org.json.JSONArray;
import java.util.*;

class RestApi {
    private final Map<String, User> users;

    RestApi(User... users) {
        this.users = new HashMap<>();
        for (User user : users) {
            this.users.put(user.name(), user);
        }
    }

    private JSONObject userToJson(User user) {
        JSONObject json = new JSONObject();
        json.put("name", user.name());

        Map<String, Double> owesMap = new HashMap<>();
        Map<String, Double> owedByMap = new HashMap<>();

        // Calculate total owes per user
        Map<String, Double> totalOwes = new HashMap<>();
        for (User.Iou iou : user.owes()) {
            totalOwes.merge(iou.getName(), iou.getAmount(), Double::sum);
        }

        // Calculate total owedBy per user
        Map<String, Double> totalOwedBy = new HashMap<>();
        for (User.Iou iou : user.owedBy()) {
            totalOwedBy.merge(iou.getName(), iou.getAmount(), Double::sum);
        }

        // Calculate net amounts
        Set<String> allUsers = new HashSet<>();
        allUsers.addAll(totalOwes.keySet());
        allUsers.addAll(totalOwedBy.keySet());

        for (String name : allUsers) {
            double owesAmount = totalOwes.getOrDefault(name, 0.0);
            double owedByAmount = totalOwedBy.getOrDefault(name, 0.0);
            double net = owesAmount - owedByAmount;

            if (net > 0) {
                owesMap.put(name, net);
            } else if (net < 0) {
                owedByMap.put(name, -net);
            }
        }

        // Calculate balance
        double balance = totalOwedBy.values().stream().mapToDouble(Double::doubleValue).sum() 
                      - totalOwes.values().stream().mapToDouble(Double::doubleValue).sum();

        json.put("owes", new JSONObject(owesMap));
        json.put("owed_by", new JSONObject(owedByMap));
        json.put("balance", Math.round(balance * 100.0) / 100.0);  // Round to 2 decimal places

        return json;
    }

    String get(String url) {
        if ("/users".equals(url)) {
            List<User> userList = new ArrayList<>(users.values());
            userList.sort(Comparator.comparing(User::name));
            JSONArray usersJson = new JSONArray();
            for (User user : userList) {
                usersJson.put(userToJson(user));
            }
            return new JSONObject().put("users", usersJson).toString();
        }
        throw new IllegalArgumentException("Unsupported URL: " + url);
    }

    String get(String url, JSONObject payload) {
        if ("/users".equals(url)) {
            JSONArray requestedNames = payload.getJSONArray("users");
            List<User> userList = new ArrayList<>();
            for (int i = 0; i < requestedNames.length(); i++) {
                String name = requestedNames.getString(i);
                User user = users.get(name);
                if (user != null) {
                    userList.add(user);
                }
            }
            userList.sort(Comparator.comparing(User::name));
            JSONArray usersJson = new JSONArray();
            for (User user : userList) {
                usersJson.put(userToJson(user));
            }
            return new JSONObject().put("users", usersJson).toString();
        }
        throw new IllegalArgumentException("Unsupported URL: " + url);
    }

    String post(String url, JSONObject payload) {
        if ("/add".equals(url)) {
            String userName = payload.getString("user");
            if (users.containsKey(userName)) {
                throw new IllegalArgumentException("User already exists: " + userName);
            }
            User newUser = User.builder().setName(userName).build();
            users.put(userName, newUser);
            return userToJson(newUser).toString();
        } else if ("/iou".equals(url)) {
            String lenderName = payload.getString("lender");
            String borrowerName = payload.getString("borrower");
            double amount = payload.getDouble("amount");

            User lender = users.get(lenderName);
            User borrower = users.get(borrowerName);
            
            if (lender == null || borrower == null) {
                throw new IllegalArgumentException("Lender or borrower not found");
            }

            // Update lender (add to their owedBy list)
            User.Builder lenderBuilder = User.builder().setName(lenderName);
            for (User.Iou iou : lender.owedBy()) {
                lenderBuilder.owedBy(iou.getName(), iou.getAmount());
            }
            lenderBuilder.owedBy(borrowerName, amount);
            for (User.Iou iou : lender.owes()) {
                lenderBuilder.owes(iou.getName(), iou.getAmount());
            }
            User newLender = lenderBuilder.build();
            users.put(lenderName, newLender);

            // Update borrower (add to their owes list)
            User.Builder borrowerBuilder = User.builder().setName(borrowerName);
            for (User.Iou iou : borrower.owes()) {
                borrowerBuilder.owes(iou.getName(), iou.getAmount());
            }
            borrowerBuilder.owes(lenderName, amount);
            for (User.Iou iou : borrower.owedBy()) {
                borrowerBuilder.owedBy(iou.getName(), iou.getAmount());
            }
            User newBorrower = borrowerBuilder.build();
            users.put(borrowerName, newBorrower);

            // Return the updated users sorted by name
            List<User> updatedUsers = Arrays.asList(newLender, newBorrower);
            updatedUsers.sort(Comparator.comparing(User::name));
            JSONArray usersJson = new JSONArray();
            for (User user : updatedUsers) {
                usersJson.put(userToJson(user));
            }
            return new JSONObject().put("users", usersJson).toString();
        }
        throw new IllegalArgumentException("Unsupported URL: " + url);
    }

}

File: src/2025-02-24-deepclaude-rust/python/exercises/practice/rest-api/rest_api.py
import json

class RestAPI:
    def __init__(self, database=None):
        self.users = {}
        if database:
            for user_data in database.get('users', []):
                name = user_data['name']
                self.users[name] = {
                    'owes': user_data.get('owes', {}),
                    'owed_by': user_data.get('owed_by', {})
                }

    def get(self, url, payload=None):
        if url == '/users':
            requested_names = []
            if payload:
                payload_data = json.loads(payload)
                requested_names = payload_data.get('users', [])
            else:
                requested_names = list(self.users.keys())
            users_list = []
            for name in sorted(requested_names):
                user = self.users.get(name)
                if user:
                    owes = {k: v for k, v in user['owes'].items()}
                    owed_by = {k: v for k, v in user['owed_by'].items()}
                    balance = sum(owed_by.values()) - sum(owes.values())
                    users_list.append({
                        'name': name,
                        'owes': owes,
                        'owed_by': owed_by,
                        'balance': balance
                    })
            users_list.sort(key=lambda x: x['name'])
            return json.dumps({'users': users_list})
        return json.dumps({})

    def post(self, url, payload=None):
        if url == '/add':
            data = json.loads(payload)
            user_name = data['user']
            if user_name not in self.users:
                self.users[user_name] = {'owes': {}, 'owed_by': {}}
            new_user = {
                'name': user_name,
                'owes': {},
                'owed_by': {},
                'balance': 0.0
            }
            return json.dumps(new_user)
        
        elif url == '/iou':
            data = json.loads(payload)
            lender = data['lender']
            borrower = data['borrower']
            amount = data['amount']
            
            lender_data = self.users.get(lender)
            borrower_data = self.users.get(borrower)
            if not lender_data or not borrower_data:
                return json.dumps({'users': []})

            # Calculate net balance between users
            existing_debt = borrower_data['owes'].get(lender, 0.0) - lender_data['owes'].get(borrower, 0.0)
            new_debt = existing_debt + amount
            
            # Clear existing entries
            if lender in borrower_data['owes']:
                del borrower_data['owes'][lender]
            if borrower in lender_data['owed_by']:
                del lender_data['owed_by'][borrower]
            if borrower in lender_data['owes']:
                del lender_data['owes'][borrower]
            if lender in borrower_data['owed_by']:
                del borrower_data['owed_by'][lender]
            
            if new_debt > 0:
                borrower_data['owes'][lender] = new_debt
                lender_data['owed_by'][borrower] = new_debt
            elif new_debt < 0:
                lender_data['owes'][borrower] = -new_debt
                borrower_data['owed_by'][lender] = -new_debt
            
            # Prepare response
            users = []
            for name in sorted([lender, borrower]):
                user = self.users[name]
                owes_total = sum(user['owes'].values())
                owed_total = sum(user['owed_by'].values())
                users.append({
                    'name': name,
                    'owes': user['owes'],
                    'owed_by': user['owed_by'],
                    'balance': owed_total - owes_total
                })
            
            return json.dumps({'users': users})
            
        return json.dumps({})
